#389 Prevent progressbar spill over new line

yasharmaster opened the pull request:
Fixes #389 
Signed-off-by: Yash Sharma yasharmaster@gmail.com

yasharmaster commented:
@pombredanne I have hard coded the allowed size of a filename to be 20 characters.
Please review the changes.

pombredanne commented on src/scancode/cli.py:
```diff
@@ -382,6 +382,9 @@ def scan_event(item):
                     return ''
                 if item:
                     _scan_success, _scanned_path = item
+                    if len(_scanned_path)>20:
```
Your spacing is not right. We use PEP8 for code standard. eg. `len(scanned_path) > 20`

pombredanne commented on src/scancode/cli.py:
```diff
@@ -382,6 +382,9 @@ def scan_event(item):
                 if item:
                     _scan_success, _scanned_path = item
+                    if len(_scanned_path)>20:
+                    	current_filename, current_file_extension = os.path.splitext(_scanned_path)
```
Please consider the use of the `commoncode.fileutils` functions instead of splitext

pombredanne commented on src/scancode/cli.py:
```diff
@@ -382,6 +382,9 @@ def scan_event(item):
                     _scan_success, _scanned_path = item
+                    if len(_scanned_path)>20:
+                    	current_filename, current_file_extension = os.path.splitext(_scanned_path)
+                    	_scanned_path = current_filename[:3] + '...' + current_filename[-3:] + current_file_extension
```
keeping only three characters at start and end does not sound right.
IMHO you should also isolate this in its own function so it can be tested.

yasharmaster commented:
@pombredanne please review the changes. I have -
[*] used `commoncode.fileutils` instead of `splitext`
[*] implemented file name shortening in a separate function
[*] set max filename size to be 25 chars, more than 25 was spilling progress bar in my system
[*] improved shortened filename by allowing more than 3 chars from beginning and the end

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,20 @@ def scan_event(item):
     return files_count, cached_scan.iterate(scans, root_dir)
 
 
+def _shorten_file_path(file_path, max_length):
```
do not use a _ prefix here.

pombredanne commented:
ping.

yasharmaster commented:
@pombred

yasharmaster commented:
I am travelling right now. Will finish these edits by tonight.

yasharmaster commented:
@pombredanne Default terminal width is 80 chars across all OS and the text before the filename spans across 55 characters. So, only 25 chars can be safely allowed for the filename if we hardcode the value. Otherwise we can do something like this https://gist.github.com/jtriley/1108174 to find current terminal width. What do you think?

yasharmaster commented:
@pombredanne please review

pombredanne commented on src/scancode/cli.py:
```diff
@@ -44,6 +44,8 @@
 import click
 from click.termui import style
 
+from math import floor
```
This is a stdlib import and should go with other stdlib imports, before `import os`

pombredanne commented on src/scancode/cli.py:
```diff
@@ -382,6 +384,7 @@ def scan_event(item):
                     return ''
                 if item:
                     _scan_success, _scanned_path = item
+                    _scanned_path = shorten_filename(_scanned_path)
```
this should only be done if we are not verbose. Note also that you are never passing a filename but a path. So your assumptions are not right in your function.

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
+	otherwise return its shortened version 
+	"""
+	if len(filename) <= max_length:
+		shortened_filename = filename
```
why not just return early?

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
+	"""
+	if len(filename) <= max_length:
+		shortened_filename = filename
+	else:
```
if you return early above, `else` is no longer needed

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
+		file_extension = fileutils.file_extension(filename)
+		number_of_dots = 3
+		remaining_length = max_length - len(file_extension) - number_of_dots
+		number_of_beginning_chars = int(floor(remaining_length/2))
```
pep8 spacing please. using abs and  a floor `//` division would  be simpler and clearer and more explicit:
`abs(remaining_length // 2)`

Also I am not sure that `number_of_beginning_chars` is a good variable name as you use it to compute both the prefix and suffix length to keep from the base name

pombredanne commented on src/scancode/cli.py:
```diff
@@ -382,6 +384,7 @@ def scan_event(item):
                 if item:
                     _scan_success, _scanned_path = item
+                    _scanned_path = shorten_filename(_scanned_path)
                     _progress_line = verbose and _scanned_path or fileutils.file_name(_scanned_path)
```
This code cannot work  correctly with the new code that you injected?

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
 
+def shorten_filename(filename, max_length=25):
+	"""
+	Return the original filename if it is short enough
```
You are receiving a path and not a name. So your docstring is misleading

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
+		shortened_filename = filename
+	else:
+		file_base_name = fileutils.file_base_name(filename)
+		file_extension = fileutils.file_extension(filename)
```
use instead  something like 
`base_name, ext = fileutils.splitext(filename)` .... but again you did not receive a filename but a path (though it does not matter to splitext)

pombredanne commented on src/scancode/cli.py:
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
+		number_of_beginning_chars = int(floor(remaining_length/2))
+		shortened_filename = (file_base_name[:number_of_beginning_chars] + 
+							number_of_dots*'.' + 
+							file_base_name[-number_of_beginning_chars:] +
```
Your pep8 spacing is not correct in this code block

pombredanne replied:
You should prefer a more explicit interpolation or format. Something like `'{start}...{end}'.format(**locals())` were you compute `start` and `end` before.

pombredanne commented on tests/scancode/data/long_file_name/0123456789012345678901234567890123456789.c:
```diff
@@ -0,0 +1,1967 @@
+/* deflate.c -- compress data using the deflation algorithm
```
Why use such a big file for tests when you only care about the name? and not the content? This is wasted space.

pombredanne commented on tests/scancode/data/long_file_name/expected.json:
```diff
@@ -0,0 +1,36 @@
+{
```
You should not care about the detailed scan results when you tests name shortening in the CLI ui.

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -169,6 +169,15 @@ def test_scan_info_does_collect_infos_with_root(monkeypatch):
+    result = runner.invoke(cli.scancode, ['--copyright', test_dir, result_file], catch_exceptions=True)
+    assert result.exit_code == 0
+    assert 'Scanning done' in result.output
+    check_scan(test_env.get_test_loc('long_file_name/expected.json'), result_file, regen=False)
```
This is not needed. We do not care about the scan. On the other hand you make no assertions on the shortening of the name in the UI, so you are not testing anything there

pombredanne commented:
@yasharmaster When you are done with updating your code, you will also need to squash your commit in one clean commit. But only then.

yasharmaster commented:
@pombredanne Please review. Should I send a new PR with this last single commit ?

pombredanne commented:
@yasharmaster nope, you can just rebase, squash and force push to that same branch of yours. This will land right here in the same PR aqlright.

yasharmaster commented:
@pombredanne Can you please point me towards the existing unit tests in the repo? Also the location where unit tests of shorten filename should go? Thanks

pombredanne commented:
Just look anywhere in the tests/ directory. There are like 1000's of them. You can add your tests alright in the same test_cli.py module
An example of a simple unit tests could be this: https://github.com/nexB/scancode-toolkit/blob/21b6c8456f6cfd42ffc07411312cc489adec26d5/tests/commoncode/test_fileutils.py#L297

yasharmaster commented:
@pombredanne Are these unit tests sufficient? I have now addressed all the comments from your previous review.

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -463,3 +464,17 @@ def test_scan_can_run_from_other_directory():
+
+class TestCLI(TestCase):
+
+	def test_shorten_filename(self):
```
You should use 4 spaces. not 8

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -463,3 +464,17 @@ def test_scan_can_run_from_other_directory():
+	def test_shorten_filename(self):
+	    assert '0123456789...0123456789.c' == cli.shorten_filename('0123456789012345678901234567890123456789.c')
+	    assert '012345.c' == cli.shorten_filename('012345.c')
+	    assert '012345678901234567890.cpp' == cli.shorten_filename('012345678901234567890.cpp')
```
you should add tests for limit args: empty, None, list, string of a length that is and a length that is not shortened right around your 25 limit.  It may help to use variables and give names to make things more explicit. Also you need to test things without extensions and full paths.
In general it is better to create one test function for one test, though the example I gave you did not follow that pattern.

yasharmaster replied:
@pombredanne what should be the output of shorten filename if a list is passed as the filename?

pombredanne replied:
what do you think? probably an exception

yasharmaster commented on src/scancode/cli.py:
```diff
@@ -383,6 +384,8 @@ def scan_event(item):
                     _scan_success, _scanned_path = item
                     _progress_line = verbose and _scanned_path or fileutils.file_name(_scanned_path)
+                    if not verbose:
+                    	_progress_line = shorten_filename(_progress_line)
```
@pombredanne The logic here is such that `_progress_line` will contain the file path only if `verbose` is set to True. In that case `shorten_filename()` will not be called at all. So argument passed to  `shorten_filename()` will always be filename only and not the file path. So why are we testing `shorten_filename()` with a file path?

pombredanne replied:
good point.

yasharmaster commented:
@pombredanne I have merged all commits into one!

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -104,6 +105,17 @@ def test_verbose_option_with_copyrights(monkeypatch):
+    result = runner.invoke(cli.scancode, ['--copyright', test_dir, result_file], catch_exceptions=True)
+    assert result.exit_code == 0
+    assert 'Scanning done' in result.output
+    assert '0123456789...0123456789.c' in result.output
```
Perfect. You are getting it.

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -452,3 +464,41 @@ def test_scan_can_run_from_other_directory():
+        test = cli.shorten_filename(filename='012345678901234567890123.c')
+        expected = '0123456789...4567890123.c'
+        assert expected == test
+
```
Use only one blank line between class methods. Two lines is only between top level module members

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -452,3 +464,41 @@ def test_scan_can_run_from_other_directory():
+        assert expected == test
+
+
+    def test_with_full_path(self):
```
Good test. You should also have a test with a non-posix path

pombredanne replied:
And tests with our without posix paths that do shortening or not

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -452,3 +464,41 @@ def test_scan_can_run_from_other_directory():
     check_scan(test_env.get_test_loc(expected_file), result_file, strip_dates=True, regen=False)
+
+
+class TestShortenFilename(TestCase):
```
You should add tests that test at the boundaries: e.g. when you have exactly the max length that will be shorted., then that minus one and that plus one. This way this would spec your function fully. You may have them already, but if this is the case, the test function names are not explicit enough

pombredanne commented on src/scancode/cli.py:
```diff
@@ -467,6 +469,26 @@ def scan_event(item):
 
+def shorten_filename(filename='', max_length=25, number_of_dots=3):
+    """
+    Return the original filename if it is short enough
```
What about rather:
`Return a truncated `filename` usable for fixed width display if `filename` is longer than `max_length`.`
Also why do we need a default for `filename=''`? Calling this function without args doe not make sense. It should just error out. Nothing special needed. `number of dots` should not be an arg. This is not something I would need to ever modify.

yasharmaster replied:
@pombredanne If we don't keep `filename=''` as default, test with empty args will fail! Should I remove the test with empty args?

pombredanne replied:
Or you can test ass an exercise that the proper exception is raised in this case.

yasharmaster commented:
@pombredanne Since `shorten_filename()` will not be called without args, therefore I have removed the test with no args. I have added new tests and have renamed them to be more explicit.

pombredanne commented:
@yasharmaster you wrote in #aboutcode IRC:
> Could anyone please let me know if should raise an exception if the function is called with no args or let it just error out? 

Just let it error out alright. My only point is that if you want to have a comprehensive coverage, then you could test that an exception is thrown in these cases. Not really needed in general but a possible useful excercise for you

yasharmaster commented:
@pombredanne please review

pombredanne commented:
This looks good to me. I am merging soon.

pombredanne commented:
as a side note, I wished we would have something based on the TTY width, such as with `blessings` instead of computing things based on a guessed fixed width.

yasharmaster commented:
@pombredanne Since the code has been merged after rebase, this can be closed as well.

pombredanne commented:
@yasharmaster closing as this was merged. Thanks++
